src/restoration-state-handler.ts

```diff
@@ -0,0 +1,252 @@
+    setCookie(`showdetails-${this.context}`, detailsState, {
+      domain: this.cookieDomain,
+      expires: this.cookieExpiration,
+      path: this.cookiePath,
+    });
+  }
+
+  private loadTileViewStateFromCookies(): CollectionDisplayMode {
+    const viewState = getCookie(`view-${this.context}`);
+    const detailsState = getCookie(`showdetails-${this.context}`);
+    if (viewState === 'tiles' || viewState === undefined) return 'grid';
+    if (detailsState === 'showdetails') return 'list-detail';
+    return 'list-compact';
+  }
+
+  private persistQueryStateToUrl(state: RestorationState) {
+    const url = new URL(window.location.href);
+    const { searchParams } = url;
+    searchParams.delete('sort');
+    searchParams.delete('query');
+    searchParams.delete('page');
+    searchParams.delete('and[]');
+    searchParams.delete('not[]');
+
+    if (state.sortParam) {
+      const prefix = state.sortParam.direction === 'desc' ? '-' : '';
+      searchParams.set('sort', `${prefix}${state.sortParam.field}`);
+    }
+
+    if (state.baseQuery) {
+      searchParams.set('query', state.baseQuery);
+    }
+
+    if (state.currentPage) {
+      if (state.currentPage > 1) {
+        searchParams.set('page', state.currentPage.toString());
+      } else {
+        searchParams.delete('page');
+      }
+    }
+
+    if (state.selectedFacets) {
+      for (const [facetName, facetValues] of Object.entries(
+        state.selectedFacets
+      )) {
+        const facetEntries = Object.entries(facetValues);
+        // eslint-disable-next-line no-continue
+        if (facetEntries.length === 0) continue;
+        for (const [key, facetState] of facetEntries) {
+          const notValue = facetState === 'hidden';
+          const paramValue = `${facetName}:${key}`;
+          if (notValue) {
+            searchParams.append('not[]', paramValue);
+          } else {
+            searchParams.append('and[]', paramValue);
+          }
+        }
+      }
+    }
+
+    if (state.dateRangeQueryClause) {
+      searchParams.append('and[]', state.dateRangeQueryClause);
+    }
+    if (state.titleQuery) {
+      searchParams.append('and[]', state.titleQuery);
+    }
+    if (state.creatorQuery) {
+      searchParams.append('and[]', state.creatorQuery);
+    }
+
+    window.history.pushState(
+      {
+        sort: state.sortParam,
+        query: state.baseQuery,
+        page: state.currentPage,
+        and: state.selectedFacets,
+        not: state.selectedFacets,
+        dateRange: state.dateRangeQueryClause,
+      },
+      '',
+      url
+    );
+  }
+
+  private loadQueryStateFromUrl(): RestorationState {
+    const url = new URL(window.location.href);
+    const pageNumber = url.searchParams.get('page');
+    const searchQuery = url.searchParams.get('query');
+    const sortQuery = url.searchParams.get('sort');
+    const facetAnds = url.searchParams.getAll('and[]');
+    const facetNots = url.searchParams.getAll('not[]');
```

Comment:
Q: do you see this set growing?

Reply:
Maybe? If we add more sort / filtering. The `and[]` and `not[]` cover most of the cases, though.